File: go/idkollen/bankid_se.go
package idkollen

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// --- Shared AgeVerification types (also used by FtnEndpoint) ---

type AgeVerificationRequest struct {
	MinAge      *int    `json:"minAge,omitempty"`
	MaxAge      *int    `json:"maxAge,omitempty"`
	RefId       *string `json:"refId,omitempty"`
	CallbackUrl *string `json:"callbackUrl,omitempty"`
	RedirectUrl *string `json:"redirectUrl,omitempty"`
}

type AgeVerificationStatus interface {
	isAgeVerificationStatus()
}

type AgeVerificationPending struct {
	Id     string  `json:"id"`
	Url    *string `json:"url"`
	MinAge *int    `json:"minAge"`
	MaxAge *int    `json:"maxAge"`
}

type AgeVerificationCompleted struct {
	Id          string `json:"id"`
	AgeVerified bool   `json:"ageVerified"`
}

type AgeVerificationFailed struct {
	Id    string `json:"id"`
	Error string `json:"error"`
}

func (*AgeVerificationPending) isAgeVerificationStatus()   {}
func (*AgeVerificationCompleted) isAgeVerificationStatus() {}
func (*AgeVerificationFailed) isAgeVerificationStatus()    {}

type ageVerificationEnvelope struct {
	Status string `json:"status"`
}

func unmarshalAgeVerificationStatus(data []byte) (AgeVerificationStatus, error) {
	var env ageVerificationEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	switch env.Status {
	case "PENDING":
		var v AgeVerificationPending
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "COMPLETED":
		var v AgeVerificationCompleted
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "FAILED":
		var v AgeVerificationFailed
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	default:
		return nil, fmt.Errorf("unknown age verification status: %s", env.Status)
	}
}

// --- BankID SE request types ---

type BankIdSeAuthRequest struct {
	Ssn            *string `json:"ssn,omitempty"`
	IpAddress      *string `json:"ipAddress,omitempty"`
	CallbackUrl    *string `json:"callbackUrl,omitempty"`
	PinRequired    *bool   `json:"pinRequired,omitempty"`
	Intent         *string `json:"intent,omitempty"`
	OrgNumber      *string `json:"orgNumber,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	RefId          *string `json:"refId,omitempty"`
}

type BankIdSePhoneAuthRequest struct {
	Ssn            string  `json:"ssn"`
	CallInitiator  string  `json:"callInitiator"`
	CallbackUrl    *string `json:"callbackUrl,omitempty"`
	PinRequired    *bool   `json:"pinRequired,omitempty"`
	Intent         *string `json:"intent,omitempty"`
	OrgNumber      *string `json:"orgNumber,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	RefId          *string `json:"refId,omitempty"`
}

type BankIdSeSignRequest struct {
	Text           string  `json:"text"`
	Ssn            *string `json:"ssn,omitempty"`
	IpAddress      *string `json:"ipAddress,omitempty"`
	CallbackUrl    *string `json:"callbackUrl,omitempty"`
	PinRequired    *bool   `json:"pinRequired,omitempty"`
	Digest         *string `json:"digest,omitempty"`
	OrgNumber      *string `json:"orgNumber,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	RefId          *string `json:"refId,omitempty"`
}

type BankIdSePhoneSignRequest struct {
	Ssn            string  `json:"ssn"`
	CallInitiator  string  `json:"callInitiator"`
	Text           string  `json:"text"`
	CallbackUrl    *string `json:"callbackUrl,omitempty"`
	PinRequired    *bool   `json:"pinRequired,omitempty"`
	Digest         *string `json:"digest,omitempty"`
	OrgNumber      *string `json:"orgNumber,omitempty"`
	RequestAddress *bool   `json:"requestAddress,omitempty"`
	RefId          *string `json:"refId,omitempty"`
}

type BankIdSeVerifyRequest struct {
	QrCode string `json:"qrCode"`
}

// --- BankID SE status types ---

type BankIdSeStatus interface {
	isBankIdSeStatus()
}

type BankIdSePhoneStatus interface {
	isBankIdSePhoneStatus()
}

type BankIdSePending struct {
	Id             string  `json:"id"`
	RefId          *string `json:"refId"`
	AutoStartToken *string `json:"autoStartToken"`
	QrStartToken   *string `json:"qrStartToken"`
	QrStartSecret  *string `json:"qrStartSecret"`
	HintCode       *string `json:"hintCode"`
}

type BankIdSePendingPhone struct {
	Id       string  `json:"id"`
	RefId    *string `json:"refId"`
	HintCode *string `json:"hintCode"`
}

type BankIdSeCompleted struct {
	Id                   string  `json:"id"`
	RefId                *string `json:"refId"`
	Ssn                  string  `json:"ssn"`
	Name                 string  `json:"name"`
	GivenName            string  `json:"givenName"`
	Surname              string  `json:"surname"`
	CertStartDate        *string `json:"certStartDate"`
	Address              *string `json:"address"`
	CompanySignatoryText *string `json:"companySignatoryText"`
}

type BankIdSeFailed struct {
	Id    string  `json:"id"`
	RefId *string `json:"refId"`
	Error string  `json:"error"`
}

func (*BankIdSePending) isBankIdSeStatus()        {}
func (*BankIdSeCompleted) isBankIdSeStatus()      {}
func (*BankIdSeFailed) isBankIdSeStatus()         {}
func (*BankIdSeCompleted) isBankIdSePhoneStatus() {}
func (*BankIdSeFailed) isBankIdSePhoneStatus()    {}
func (*BankIdSePendingPhone) isBankIdSePhoneStatus() {}

type BankIdSeVerifyResponse struct {
	Ssn        string  `json:"ssn"`
	Name       string  `json:"name"`
	GivenName  string  `json:"givenName"`
	Surname    string  `json:"surname"`
	Age        *int    `json:"age"`
	VerifiedAt *string `json:"verifiedAt"`
}

type bankIdSeStatusEnvelope struct {
	Status string `json:"status"`
}

func unmarshalBankIdSeStatus(data []byte) (BankIdSeStatus, error) {
	var env bankIdSeStatusEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	switch env.Status {
	case "PENDING":
		var v BankIdSePending
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "COMPLETED":
		var v BankIdSeCompleted
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "FAILED":
		var v BankIdSeFailed
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	default:
		return nil, fmt.Errorf("unknown bankid-se status: %s", env.Status)
	}
}

func unmarshalBankIdSePhoneStatus(data []byte) (BankIdSePhoneStatus, error) {
	var env bankIdSeStatusEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	switch env.Status {
	case "PENDING":
		var v BankIdSePendingPhone
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "COMPLETED":
		var v BankIdSeCompleted
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "FAILED":
		var v BankIdSeFailed
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	default:
		return nil, fmt.Errorf("unknown bankid-se phone status: %s", env.Status)
	}
}

// --- BankID SE endpoint ---

type BankIdSeEndpoint struct {
	client *IdkollenClient
}

func (e *BankIdSeEndpoint) Auth(ctx context.Context, req BankIdSeAuthRequest) (BankIdSeStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/bankid-se/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalBankIdSeStatus(raw)
}

func (e *BankIdSeEndpoint) PhoneAuth(ctx context.Context, req BankIdSePhoneAuthRequest) (BankIdSePhoneStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/bankid-se/phone/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalBankIdSePhoneStatus(raw)
}

func (e *BankIdSeEndpoint) Sign(ctx context.Context, req BankIdSeSignRequest) (BankIdSeStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/bankid-se/sign", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalBankIdSeStatus(raw)
}

func (e *BankIdSeEndpoint) PhoneSign(ctx context.Context, req BankIdSePhoneSignRequest) (BankIdSePhoneStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/bankid-se/phone/sign", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalBankIdSePhoneStatus(raw)
}

func (e *BankIdSeEndpoint) Verify(ctx context.Context, req BankIdSeVerifyRequest) (*BankIdSeVerifyResponse, error) {
	var out BankIdSeVerifyResponse
	if err := e.client.post(ctx, "/v3/bankid-se/verify", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (e *BankIdSeEndpoint) AgeVerification(ctx context.Context, req AgeVerificationRequest) (AgeVerificationStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/bankid-se/age-verification", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalAgeVerificationStatus(raw)
}

func (e *BankIdSeEndpoint) AuthStatus(ctx context.Context, id string) (BankIdSeStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/bankid-se/auth/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalBankIdSeStatus(raw)
}

func (e *BankIdSeEndpoint) SignStatus(ctx context.Context, id string) (BankIdSeStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/bankid-se/sign/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalBankIdSeStatus(raw)
}

func (e *BankIdSeEndpoint) AgeVerificationStatus(ctx context.Context, id string) (AgeVerificationStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/bankid-se/age-verification/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalAgeVerificationStatus(raw)
}

func (e *BankIdSeEndpoint) CancelAuth(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/bankid-se/auth/"+id)
}

func (e *BankIdSeEndpoint) CancelSign(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/bankid-se/sign/"+id)
}

func (e *BankIdSeEndpoint) CancelAgeVerification(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/bankid-se/age-verification/"+id)
}

func (e *BankIdSeEndpoint) WaitForAuth(ctx context.Context, id string, opts PollOptions) (BankIdSeStatus, error) {
	return pollBankIdSe(ctx, opts, func() (BankIdSeStatus, error) { return e.AuthStatus(ctx, id) })
}

func (e *BankIdSeEndpoint) WaitForSign(ctx context.Context, id string, opts PollOptions) (BankIdSeStatus, error) {
	return pollBankIdSe(ctx, opts, func() (BankIdSeStatus, error) { return e.SignStatus(ctx, id) })
}

func (e *BankIdSeEndpoint) WaitForAgeVerification(ctx context.Context, id string, opts PollOptions) (AgeVerificationStatus, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = 2 * time.Second
	}
	for {
		status, err := e.AgeVerificationStatus(ctx, id)
		if err != nil {
			return nil, &WaitError{Err: err}
		}
		if _, ok := status.(*AgeVerificationPending); !ok {
			return status, nil
		}
		select {
		case <-ctx.Done():
			return nil, &WaitError{Timeout: true}
		case <-time.After(interval):
		}
	}
}

func pollBankIdSe(ctx context.Context, opts PollOptions, fn func() (BankIdSeStatus, error)) (BankIdSeStatus, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = 2 * time.Second
	}
	for {
		status, err := fn()
		if err != nil {
			return nil, &WaitError{Err: err}
		}
		if _, ok := status.(*BankIdSePending); !ok {
			return status, nil
		}
		select {
		case <-ctx.Done():
			return nil, &WaitError{Timeout: true}
		case <-time.After(interval):
		}
	}
}

File: ruby/lib/idkollen/bank_id_se.rb
# frozen_string_literal: true

module Idkollen
  module BankIdSe
    # --- Requests ---

    class AuthRequest
      attr_reader :ssn, :ip_address, :callback_url, :pin_required, :intent,
                  :org_number, :request_address, :ref_id

      def initialize(ssn: nil, ip_address: nil, callback_url: nil, pin_required: nil,
                     intent: nil, org_number: nil, request_address: nil, ref_id: nil)
        @ssn = ssn
        @ip_address = ip_address
        @callback_url = callback_url
        @pin_required = pin_required
        @intent = intent
        @org_number = org_number
        @request_address = request_address
        @ref_id = ref_id
      end

      def to_h
        {
          ssn: @ssn, ipAddress: @ip_address, callbackUrl: @callback_url,
          pinRequired: @pin_required, intent: @intent, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class PhoneAuthRequest
      attr_reader :ssn, :call_initiator, :callback_url, :pin_required, :intent,
                  :org_number, :request_address, :ref_id

      def initialize(ssn:, call_initiator:, callback_url: nil, pin_required: nil,
                     intent: nil, org_number: nil, request_address: nil, ref_id: nil)
        @ssn = ssn
        @call_initiator = call_initiator
        @callback_url = callback_url
        @pin_required = pin_required
        @intent = intent
        @org_number = org_number
        @request_address = request_address
        @ref_id = ref_id
      end

      def to_h
        {
          ssn: @ssn, callInitiator: @call_initiator, callbackUrl: @callback_url,
          pinRequired: @pin_required, intent: @intent, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class SignRequest
      attr_reader :text, :ssn, :ip_address, :callback_url, :pin_required, :digest,
                  :org_number, :request_address, :ref_id

      def initialize(text:, ssn: nil, ip_address: nil, callback_url: nil, pin_required: nil,
                     digest: nil, org_number: nil, request_address: nil, ref_id: nil)
        @text = text; @ssn = ssn; @ip_address = ip_address; @callback_url = callback_url
        @pin_required = pin_required; @digest = digest; @org_number = org_number
        @request_address = request_address; @ref_id = ref_id
      end

      def to_h
        {
          text: @text, ssn: @ssn, ipAddress: @ip_address, callbackUrl: @callback_url,
          pinRequired: @pin_required, digest: @digest, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class PhoneSignRequest
      attr_reader :ssn, :call_initiator, :text, :callback_url, :pin_required, :digest,
                  :org_number, :request_address, :ref_id

      def initialize(ssn:, call_initiator:, text:, callback_url: nil, pin_required: nil,
                     digest: nil, org_number: nil, request_address: nil, ref_id: nil)
        @ssn = ssn; @call_initiator = call_initiator; @text = text
        @callback_url = callback_url; @pin_required = pin_required; @digest = digest
        @org_number = org_number; @request_address = request_address; @ref_id = ref_id
      end

      def to_h
        {
          ssn: @ssn, callInitiator: @call_initiator, text: @text,
          callbackUrl: @callback_url, pinRequired: @pin_required, digest: @digest,
          orgNumber: @org_number, requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class VerifyRequest
      attr_reader :qr_code
      def initialize(qr_code:) = @qr_code = qr_code
      def to_h = { qrCode: @qr_code }
    end

    class VerifyResponse
      attr_reader :ssn, :name, :given_name, :surname, :age, :verified_at

      def initialize(ssn:, name:, given_name:, surname:, age: nil, verified_at: nil)
        @ssn = ssn; @name = name; @given_name = given_name; @surname = surname
        @age = age; @verified_at = verified_at
      end

      def deconstruct_keys(_)
        { ssn: @ssn, name: @name, given_name: @given_name, surname: @surname,
          age: @age, verified_at: @verified_at }
      end

      def self.from_json(d)
        new(ssn: d['ssn'], name: d['name'], given_name: d['givenName'],
            surname: d['surname'], age: d['age'], verified_at: d['verifiedAt'])
      end
    end

    # --- Status variants ---

    class Pending
      attr_reader :id, :ref_id, :auto_start_token, :qr_start_token, :qr_start_secret, :hint_code

      def initialize(id:, ref_id: nil, auto_start_token: nil, qr_start_token: nil,
                     qr_start_secret: nil, hint_code: nil)
        @id = id; @ref_id = ref_id; @auto_start_token = auto_start_token
        @qr_start_token = qr_start_token; @qr_start_secret = qr_start_secret
        @hint_code = hint_code
      end

      def deconstruct_keys(_)
        { id: @id, ref_id: @ref_id, auto_start_token: @auto_start_token,
          qr_start_token: @qr_start_token, qr_start_secret: @qr_start_secret,
          hint_code: @hint_code }
      end

      def self.from_json(d)
        new(id: d['id'], ref_id: d['refId'], auto_start_token: d['autoStartToken'],
            qr_start_token: d['qrStartToken'], qr_start_secret: d['qrStartSecret'],
            hint_code: d['hintCode'])
      end
    end

    class PendingPhone
      attr_reader :id, :ref_id, :hint_code

      def initialize(id:, ref_id: nil, hint_code: nil)
        @id = id; @ref_id = ref_id; @hint_code = hint_code
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, hint_code: @hint_code }

      def self.from_json(d)
        new(id: d['id'], ref_id: d['refId'], hint_code: d['hintCode'])
      end
    end

    class Completed
      attr_reader :id, :ref_id, :ssn, :name, :given_name, :surname,
                  :cert_start_date, :address, :company_signatory_text

      def initialize(id:, ssn:, name:, given_name:, surname:,
                     ref_id: nil, cert_start_date: nil, address: nil, company_signatory_text: nil)
        @id = id; @ref_id = ref_id; @ssn = ssn; @name = name
        @given_name = given_name; @surname = surname
        @cert_start_date = cert_start_date; @address = address
        @company_signatory_text = company_signatory_text
      end

      def deconstruct_keys(_)
        { id: @id, ref_id: @ref_id, ssn: @ssn, name: @name, given_name: @given_name,
          surname: @surname, cert_start_date: @cert_start_date, address: @address,
          company_signatory_text: @company_signatory_text }
      end

      def self.from_json(d)
        new(id: d['id'], ref_id: d['refId'], ssn: d['ssn'], name: d['name'],
            given_name: d['givenName'], surname: d['surname'],
            cert_start_date: d['certStartDate'], address: d['address'],
            company_signatory_text: d['companySignatoryText'])
      end
    end

    class Failed
      attr_reader :id, :ref_id, :error

      def initialize(id:, error:, ref_id: nil)
        @id = id; @ref_id = ref_id; @error = error
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, error: @error }

      def self.from_json(d) = new(id: d['id'], ref_id: d['refId'], error: d['error'])
    end

    module Status
      def self.from_json(d)
        case d['status']
        when 'PENDING'   then Pending.from_json(d)
        when 'COMPLETED' then Completed.from_json(d)
        when 'FAILED'    then Failed.from_json(d)
        else raise ArgumentError, "Unknown bankid-se status: #{d['status']}"
        end
      end
    end

    module PhoneStatus
      def self.from_json(d)
        case d['status']
        when 'PENDING'   then PendingPhone.from_json(d)
        when 'COMPLETED' then Completed.from_json(d)
        when 'FAILED'    then Failed.from_json(d)
        else raise ArgumentError, "Unknown bankid-se phone status: #{d['status']}"
        end
      end
    end

    # --- Endpoint ---

    class Endpoint
      def initialize(transport)
        @transport = transport
      end

      def auth(req)        = Status.from_json(@transport.post('/v3/bankid-se/auth', req.to_h))
      def phone_auth(req)  = PhoneStatus.from_json(@transport.post('/v3/bankid-se/phone/auth', req.to_h))
      def sign(req)        = Status.from_json(@transport.post('/v3/bankid-se/sign', req.to_h))
      def phone_sign(req)  = PhoneStatus.from_json(@transport.post('/v3/bankid-se/phone/sign', req.to_h))
      def verify(req)      = VerifyResponse.from_json(@transport.post('/v3/bankid-se/verify', req.to_h))

      def age_verification(req)
        Idkollen::AgeVerification::Status.from_json(@transport.post('/v3/bankid-se/age-verification', req.to_h))
      end

      def auth_status(id) = Status.from_json(@transport.get("/v3/bankid-se/auth/#{id}"))
      def sign_status(id) = Status.from_json(@transport.get("/v3/bankid-se/sign/#{id}"))

      def age_verification_status(id)
        Idkollen::AgeVerification::Status.from_json(@transport.get("/v3/bankid-se/age-verification/#{id}"))
      end

      def cancel_auth(id)             = @transport.delete("/v3/bankid-se/auth/#{id}")
      def cancel_sign(id)             = @transport.delete("/v3/bankid-se/sign/#{id}")
      def cancel_age_verification(id) = @transport.delete("/v3/bankid-se/age-verification/#{id}")

      def wait_for_auth(id, opts: PollOptions.new)
        poll_status(opts) { auth_status(id) }
      end

      def wait_for_sign(id, opts: PollOptions.new)
        poll_status(opts) { sign_status(id) }
      end

      def wait_for_age_verification(id, opts: PollOptions.new)
        poll_age(opts) { age_verification_status(id) }
      end

      private

      def poll_status(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end

      def poll_age(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Idkollen::AgeVerification::Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end
    end
  end
end
